src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -5,73 +5,223 @@
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
+
+        public override Task<SemanticTokens> GetSemanticTokensAsync(DocumentSnapshot documentSnapshot, TextDocumentIdentifier textDocumentIdentifier, CancellationToken cancellationToken)
         {
-            return GetSemanticTokens(codeDocument, range: null);
+            return GetSemanticTokensAsync(documentSnapshot, textDocumentIdentifier, range: null, cancellationToken);
         }
 
-        public override SemanticTokens GetSemanticTokens(RazorCodeDocument codeDocument, Range range)
+        public override async Task<SemanticTokens> GetSemanticTokensAsync(
+            DocumentSnapshot documentSnapshot,
+            TextDocumentIdentifier textDocumentIdentifier,
+            Range range,
+            CancellationToken cancellationToken)
         {
+            var codeDocument = await GetRazorCodeDocument(documentSnapshot);
             if (codeDocument is null)
             {
                 throw new ArgumentNullException(nameof(codeDocument));
             }
+            var documentVersion = await documentSnapshot.GetTextVersionAsync();
 
-            var semanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range);
+            var razorSemanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument, range);
+            var csharpSemanticRanges = await GetCSharpSemanticRangesAsync(codeDocument, textDocumentIdentifier, range, cancellationToken);
+            var combinedSemanticRanges = CombineSemanticRanges(razorSemanticRanges, csharpSemanticRanges);
 
-            var semanticTokens = ConvertSemanticRangesToSemanticTokens(semanticRanges, codeDocument);
+            var razorSemanticTokens = ConvertSemanticRangesToSemanticTokens(combinedSemanticRanges, codeDocument, documentVersion);
 
-            return semanticTokens;
+            return razorSemanticTokens;
         }
 
-        public override SemanticTokensFullOrDelta GetSemanticTokensEdits(RazorCodeDocument codeDocument, string previousResultId)
+        public override async Task<SemanticTokensFullOrDelta> GetSemanticTokensEditsAsync(
+            DocumentSnapshot documentSnapshot,
+            TextDocumentIdentifier textDocumentIdentifier,
+            string previousResultId,
+            CancellationToken cancellationToken)
         {
+            var codeDocument = await GetRazorCodeDocument(documentSnapshot);
             if (codeDocument is null)
             {
                 throw new ArgumentNullException(nameof(codeDocument));
             }
+            var documentVersion = await documentSnapshot.GetTextVersionAsync();
 
-            var semanticRanges = TagHelperSemanticRangeVisitor.VisitAllNodes(codeDocument);
-
+            VersionStamp? cachedVersionStamp = null;
             IReadOnlyList<int> previousResults = null;
+            if(previousResultId != null)
```

Comment:
Should do a format on this doc
```suggestion
            if (previousResultId != null)
```

Reply:
Wish VS had a setting to just format on save.